File: src/main/java/org/tdf/lotusvm/types/ResultType.java
package org.tdf.lotusvm.types;


import org.tdf.lotusvm.common.BytesReader;

import static org.tdf.lotusvm.common.Constants.*;

public enum ResultType {
    EMPTY(RESULT_EMPTY),
    I32(VALUE_I32),
    I64(VALUE_I64),
    F32(VALUE_F32),
    F64(VALUE_F64);
    public final int code;

    ResultType(int code) {
        this.code = code;
    }

    static ResultType readFrom(BytesReader reader) {
        int type = reader.read();
        switch (type) {
            case VALUE_I32:
                return I32;
            case VALUE_I64:
                return I64;
            case VALUE_F32:
                return F32;
            case VALUE_F64:
                return F64;
            case RESULT_EMPTY:
                return EMPTY;
            default:
                throw new IllegalArgumentException(String.format("unknown value type %x", type));
        }
    }
}

File: src/main/java/org/tdf/lotusvm/types/ResultType.kt
package org.tdf.lotusvm.types

import org.tdf.lotusvm.common.BytesReader
import org.tdf.lotusvm.common.Constants

enum class ResultType(val code: Int) {
    EMPTY(Constants.RESULT_EMPTY), I32(Constants.VALUE_I32), I64(Constants.VALUE_I64), F32(Constants.VALUE_F32), F64(
        Constants.VALUE_F64
    );

    companion object {
        @JvmField
        val VALUES = values()

        @JvmStatic
        fun readFrom(reader: BytesReader): ResultType {
            val type = reader.read()
            return when (type) {
                Constants.VALUE_I32 -> I32
                Constants.VALUE_I64 -> I64
                Constants.VALUE_F32 -> F32
                Constants.VALUE_F64 -> F64
                Constants.RESULT_EMPTY -> EMPTY
                else -> throw IllegalArgumentException(String.format("unknown value type %x", type))
            }
        }
    }
}